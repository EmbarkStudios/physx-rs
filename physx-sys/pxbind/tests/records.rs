struct RecordOutput {
    structgen: String,
    size_asserts: String,
    rust_decls: String,
}

fn gen_records(which: &str, to_emit: &'static [&str]) -> anyhow::Result<RecordOutput> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/record/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |rb: &pxbind::consumer::RecBinding<'_>| {
        to_emit.is_empty() || to_emit.iter().any(|te| *te == rb.name)
    };

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        ..Default::default()
    };

    let structgen = {
        let mut sg = Vec::new();
        generator.generate_structgen(&consumer, &mut sg)?;
        String::from_utf8(sg)?
    };

    let size_asserts = {
        let mut sa = Vec::new();
        generator.generate_size_asserts(&consumer, &mut sa)?;
        String::from_utf8(sa)?
    };

    let rust_decls = {
        let mut rd = Vec::new();
        generator.generate_rust_records(&consumer, &mut rd)?;
        String::from_utf8(rd)?
    };

    Ok(RecordOutput {
        structgen,
        size_asserts,
        rust_decls,
    })
}

/// Verifies that abstract base classes with no fields are properly bound
#[test]
fn abstract_() {
    let ro = gen_records("abstract.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Verifies that reference fields are reflected in the size and layout, but aren't
/// actually publicly exposed
#[test]
fn ref_fields() {
    let ro = gen_records("ref_fields.h", &["PxReffyMcRefface"]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensures we can generate PODs that wrap types that are only exposed publicly
/// via pointer
#[test]
fn ptr_only() {
    let ro = gen_records("ptr_only.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensures we can generate PODs for some templates types that are unfortunately
/// now part of the core API
#[test]
fn simple_templates() {
    let ro = gen_records("simple_templates.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensure we can parse templates, template specializations, and template parameters
/// and stamp out unique types for each template + template param
#[test]
fn templates() {
    let ro = gen_records(
        "templates.h",
        &[
            "PxRaycastCallback",
            "PxOverlapCallback",
            "PxSweepCallback",
            "PxRaycastBuffer",
            "PxOverlapBuffer",
        ],
    )
    .unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// More complicated records with multiple levels of inheritance
#[test]
fn complex() {
    let ro = gen_records("complex.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}
