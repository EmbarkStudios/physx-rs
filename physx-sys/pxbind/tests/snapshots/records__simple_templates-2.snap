---
source: physx-sys/pxbind/tests/records.rs
expression: ro.size_asserts
---
using namespace physx;
#include "structgen_out.hpp"

static_assert(sizeof(physx::PxVec3) == sizeof(physx_PxVec3_Pod), "POD wrapper for `physx::PxVec3` has incorrect size
static_assert(sizeof(physx::PxVec3Padded) == sizeof(physx_PxVec3Padded_Pod), "POD wrapper for `physx::PxVec3Padded` has incorrect size
static_assert(sizeof(physx::PxVec4) == sizeof(physx_PxVec4_Pod), "POD wrapper for `physx::PxVec4` has incorrect size
static_assert(sizeof(physx::PxMat44) == sizeof(physx_PxMat44_Pod), "POD wrapper for `physx::PxMat44` has incorrect size
static_assert(sizeof(physx::PxQuat) == sizeof(physx_PxQuat_Pod), "POD wrapper for `physx::PxQuat` has incorrect size


